File: Easy/1598. Crawler Log Folder/solution1.js
/**
 * @param {string[]} logs
 * @return {number}
 */
const minOperations = function (logs) {
    let result = 0;

    for(let i = 0; i<logs.length; i++) {
        if(logs[i] === "../") {
            result -= result !== 0;
        } else if(logs[i] === "./") {
            continue;
        } else {
            result++;
        }
    }

    return result;
};


// ------------------------
console.log(minOperations(["d1/","d2/","../","d21/","./"]));
console.log(minOperations(["d1/","d2/","./","d3/","../","d31/"]));
console.log(minOperations(["d1/","../","../","../"]));

File: Easy/1598. Crawler Log Folder/solution1.rs
impl Solution {
    pub fn min_operations(logs: Vec<String>) -> i32 {
        let mut result: i32 = 0;

        for log in logs.iter() {
            match &log[..] {
                "../" => result -= if result == 0 { 0 } else { 1 },
                "./" => continue,
                _ => result += 1
            }
        }

        return result;
    }
}

// --------------------
struct Solution;

// Ref: https://stackoverflow.com/a/38183903/8595750
macro_rules! vec_of_strings {
    ($($x:expr),*) => (vec![$($x.to_string()),*]);
}

fn main() {
    println!(
        "{}",
        Solution::min_operations(vec_of_strings!["d1/", "d2/", "../", "d21/", "./"])
    );
    println!(
        "{}",
        Solution::min_operations(vec_of_strings!["d1/","d2/","./","d3/","../","d31/"])
    );
    println!(
        "{}",
        Solution::min_operations(vec_of_strings!["d1/","../","../","../"])
    );
}
